ui/components/app/assets/util/filter.ts

```diff
@@ -0,0 +1,67 @@
+import { get } from 'lodash';
+
+export type FilterCriteria = {
+  key: string;
+  opts?: Record<string, FilterType>; // Use opts for range, inclusion, etc.
+  filterCallback?: FilterCallbackKeys; // Specify the type of filter: 'range', 'inclusive', etc.
+};
+
+export type FilterType = string | number | boolean | Date;
+type FilterCallbackKeys = keyof FilterCallbacksT;
+
+export type FilterCallbacksT = {
+  inclusive: (value: string, opts: Record<string, boolean>) => boolean;
+  range: (value: number, opts: Record<string, number>) => boolean;
+};
+
+const filterCallbacks: FilterCallbacksT = {
+  inclusive: (value: string, opts: Record<string, boolean>) => opts[value],
+  range: (value: number, opts: Record<string, number>) =>
+    value >= opts.min && value <= opts.max,
+};
+
+function getNestedValue<T>(obj: T, keyPath: string): FilterType {
+  return get(obj, keyPath);
+}
+
+export function filterAssets<T>(array: T[], criteria: FilterCriteria[]): T[] {
+  if (criteria.length === 0) {
+    return array;
+  }
+
+  return array.filter((item) => {
+    return criteria.every((criterion) => {
+      const { key, opts, filterCallback } = criterion;
+      const nestedValue = getNestedValue(item, key);
+
+      if (filterCallback && opts) {
+        switch (filterCallback) {
+          case 'inclusive':
+            if (typeof nestedValue === 'string') {
```

Comment:
what if you create helper functions to handle `inclusive` filter and `range` filter like:
```
// Helper function for 'inclusive' filter
function handleInclusiveFilter(nestedValue: unknown, opts: Record<string, boolean>): boolean {
  return typeof nestedValue === 'string'
    ? filterCallbacks.inclusive(nestedValue, opts)
    : false;
}

// Helper function for 'range' filter
function handleRangeFilter(nestedValue: unknown, opts: { min: number; max: number }): boolean {
  return typeof nestedValue === 'number'
    ? filterCallbacks.range(nestedValue, opts)
    : false;
}
```
so that you can then have something much more readable like:
```
switch (filterCallback) {
        case 'inclusive':
          return handleInclusiveFilter(nestedValue, opts);
        case 'range':
          return handleRangeFilter(nestedValue, opts);
        default:
          return true;
      }
```

Reply:
Yeah you're right this could be cleaner. Rather than adding the additional abstraction, I ended up just cleaning up the code in the switch statement, would you be fine with this?

```
      switch (filterCallback) {
        case 'inclusive':
          return filterCallbacks.inclusive(
            nestedValue as string,
            opts as Record<string, boolean>,
          );
        case 'range':
          return filterCallbacks.range(
            nestedValue as number,
            opts as { min: number; max: number },
          );
        default:
          return true;
      }
 ```